File: submissions/cpp/79.cpp
class Solution {
public:
    bool exist(vector<vector<char>> &board, string word) {
        int n = board.size();
        if (n == 0)
            return false;
        int m = board[0].size(), len = word.size();
        if (n == 1 && m == 1 && len == 1 && board[0][0] == word[0])
            return true;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (board[i][j] == word[0]) {
                    char tmp = board[i][j];
                    board[i][j] = \u0027\u005C0\u0027;
                    SearchWord(board, word, n, m, i - 1, j, 1, len);
                    SearchWord(board, word, n, m, i + 1, j, 1, len);
                    SearchWord(board, word, n, m, i, j - 1, 1, len);
                    SearchWord(board, word, n, m, i, j + 1, 1, len);
                    board[i][j] = tmp;
                }
            }
        }
        return found;
    }

    void SearchWord(vector<vector<char>> &board, string word, int &n, int &m, int i, int j, int k, int &len) {
        if (found)
            return;
        if (i < 0 || j < 0 || i >= n || j >= m)
            return;
        if (k == len) {
            found = true;
            return;
        }
        if (board[i][j] == word[k]) {
            char tmp = board[i][j];
            board[i][j] = \u0027\u005C0\u0027;
            SearchWord(board, word, n, m, i - 1, j, k + 1, len);
            SearchWord(board, word, n, m, i + 1, j, k + 1, len);
            SearchWord(board, word, n, m, i, j - 1, k + 1, len);
            SearchWord(board, word, n, m, i, j + 1, k + 1, len);
            board[i][j] = tmp;
        }
    }

private:
    bool found = false;
};

File: LeetCode/python3/79.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0]) if len(board) > 0 else 0
        if m == 0:
            return False
        if len(word) == 0:
            return True
        dir_i, dir_j = [0, 1, 0, -1], [1, 0, -1, 0]

        def BackTrack(i, j, k):
            if k == len(word):
                return True
            temp = board[i][j]
            board[i][j] = '!'
            for l in range(4):
                new_i, new_j = i + dir_i[l], j + dir_j[l]
                if new_i >= 0 and new_i < m and new_j >= 0 and new_j < n and board[new_i][new_j] == word[k]:
                    if BackTrack(new_i, new_j, k + 1):
                        return True
            board[i][j] = temp

        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:
                    if BackTrack(i, j, 1):
                        return True
        return False
